src/ace/sparse_ace_utils.jl

```diff
@@ -146,19 +146,26 @@ end
 """
 takes an nnll spec and generates a complete list of all possible nnllmm
 """
-function _auto_nnllmm_spec(nnll_spec)
+function _auto_nnllmm_spec(nnll_spec, Lmax = Inf)
+   # NOTE: this function is a huge bottleneck of the basis generation code 
+   #       but it appears that it cannot be easily improved. Using sorted 
+   #       inserts is MUCH MUCH slower. Using a Set is also a little bit 
+   #       slower. 
+   #       Consider how to multi-thread it? 
    _sortby(bb) = (length(bb), bb) 
-
    NT_NLM = typeof( (n = 0, l = 0, m = 0) ) 
    nnllmm = Vector{NT_NLM}[] 
    for bb in nnll_spec
       MM = setproduct( [ -b.l:b.l for b in bb ] )
       for mm in eachrow(MM)
-         push!(nnllmm, [ (n = b.n, l = b.l, m = m) 
-                         for (b, m) in zip(bb, mm) ] )
+         # Liwei: is this really correct? 
+         if abs(sum(mm)) > Lmax; continue; end 
```

Comment:
@zhanglw0521  --  is this line really correct? My understanding is that it should be ok for complex but not for real couplings. It feels like line 78 in `symmop.jl` 
```julia
         _i = (inv_𝔸spec[ (nn, ll, mm) ])::Int 
```
should fail or return garbage...

But somehow the tests pass anyhow? How can they?

Reply:
ok - scratch this, the examples do fail now. It seems our tests are just not comprehensive enough. Weird.